src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveModuleBuilderTests.cs

```diff
@@ -0,0 +1,262 @@
+            AssertExtensions.Throws<ArgumentException>("name", () => module.DefineUninitializedData("", 1, FieldAttributes.Public));
+
+            foreach (int size in new int[] { -1, 0, 0x003f0000, 0x003f0000 + 1 })
+            {
+                AssertExtensions.Throws<ArgumentException>("size", () => module.DefineUninitializedData("TestField", size, FieldAttributes.Private));
+            }
+
+            module.CreateGlobalFunctions();
+            Assert.Throws<InvalidOperationException>(() => module.DefineUninitializedData("TestField", 1, FieldAttributes.HasDefault));
+        }
+
+        [Theory]
+        [InlineData(FieldAttributes.Static | FieldAttributes.Public)]
+        [InlineData(FieldAttributes.Static | FieldAttributes.Private)]
+        [InlineData(FieldAttributes.Private)]
+        public void DefineInitializedDataTest(FieldAttributes attributes)
+        {
+            AssemblyBuilder ab = AssemblySaveTools.PopulateAssemblyBuilderAndSaveMethod(new AssemblyName("MyAssembly"), out MethodInfo _);
+            ModuleBuilder module = ab.DefineDynamicModule("MyModule");
+            FieldBuilder field = module.DefineInitializedData("MyField", [01, 00, 01], attributes);
+
+            Assert.True(field.IsStatic);
+            Assert.Equal((attributes & FieldAttributes.Public) != 0, field.IsPublic);
+            Assert.Equal((attributes & FieldAttributes.Private) != 0, field.IsPrivate);
+            Assert.Equal("MyField", field.Name);
+        }
+
+        [Fact]
+        public void DefineInitializedData_Validations()
+        {
+            AssemblyBuilder ab = AssemblySaveTools.PopulateAssemblyBuilderAndSaveMethod(new AssemblyName("MyAssembly"), out MethodInfo _);
+            ModuleBuilder module = ab.DefineDynamicModule("MyModule");
+
+            AssertExtensions.Throws<ArgumentNullException>("name", () => module.DefineInitializedData(null, [1, 0, 1], FieldAttributes.Public));
+            AssertExtensions.Throws<ArgumentException>("name", () => module.DefineInitializedData("", [1, 0, 1], FieldAttributes.Private));
+            AssertExtensions.Throws<ArgumentNullException>("data", () => module.DefineInitializedData("MyField", null, FieldAttributes.Public));
+            AssertExtensions.Throws<ArgumentException>("size", () => module.DefineInitializedData("MyField", [], FieldAttributes.Public));
+            AssertExtensions.Throws<ArgumentException>("size", () => module.DefineInitializedData("MyField", new byte[0x3f0000], FieldAttributes.Public));
+
+            FieldBuilder field = module.DefineInitializedData("MyField", [1, 0, 1], FieldAttributes.Public);
+            module.CreateGlobalFunctions();
+
+            Assert.Null(field.DeclaringType);
+            Assert.Throws<InvalidOperationException>(() => module.DefineInitializedData("MyField2", new byte[] { 1, 0, 1 }, FieldAttributes.Public));
+        }
+
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotMonoRuntime))]
```

Comment:
The Assert at row 257 fails on mono, not sure it is expected ...
```cs
Assert.Equal(name + "_0" + "_" + address.ToString(), name + "_" + (address % minAlignmentRequired).ToString() + "_" + address.ToString());
```

Reply:
I can see the failure locally after building with mono runtime, though could not find the root cause, so filed an [issue](https://github.com/dotnet/runtime/issues/97172).

Reply:
@buyaa-n I don't understand the test.  What is forcing the address to be aligned? is it just based on the length of the blob of data?

I'm looking at sections II.16.3.1 ("Data declaration") and II.16.3.2 ("Accessing data from the PE file") in ECMA-335 and I don't see anything about alignment.

The only mention of alignment I see is:

> (Each data item shall have a DataLabel if it is to be referenced afterwards; missing a DataLabel is useful in order to insert alignment padding between data items)

so is `DefineInitializedData` responsible for adding the padding?  And the test is verifying that the right amount got added?

Reply:
> I don't understand the test. What is forcing the address to be aligned? is it just based on the length of the blob of data?
> 
> I'm looking at sections II.16.3.1 ("Data declaration") and II.16.3.2 ("Accessing data from the PE file") in ECMA-335 and I don't see anything about alignment.

Right, when I encountered below assert that is failing when a byte array with a size of 1 or 4 (not multiple of 8) is provided for the field RVA data, I also checked ECMA spec and did not found anything about this alignment requirement 
https://github.com/dotnet/runtime/blob/2f63046c28f2f57f2c8666845f1c1de8696b1626/src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/ManagedTextSection.cs#L190
It looks like a runtime requirement, not ECMA spec, anyway I had to add alignment for each field RVA: https://github.com/dotnet/runtime/blob/f36b7cded294d07962060856843c5fa8ec1beb7d/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs#L453-L454
Then this test and the metadata Assert passes.

> so is DefineInitializedData responsible for adding the padding? And the test is verifying that the right amount got added?

By my understanding yes, the failing part of the test is asserting if the fields aligned properly.  Before adding the alignment same failure was happening locally, now this fail on mono even there is no mono specific code